### PR TITLE
Upgrade to Node 20.12.2

### PR DESCRIPTION
**Pull request by joshuacurtiss**
In addition to upgrading the project to Node 20.12.2, including its GitHub Actions job, we also made these adjustments:

When regenerating the `package-lock.json` file, `browserslist` made updates that cause Babel processing to *allow* `const` in the transformed output. (This is because the latest browser data eliminates `safari 9.1` from `browserslist` with our settings, and Babel thus decided to keep `const` in its output.) As a result, uglification was throwing errors:

```
Warning: Uglification failed.
Unexpected token: keyword (const).
```

Upgrading `grunt-contrib-uglify` to v4.0.1 resolves the uglification error, but we use the latest v5.2.2 as a consistent version across projects.

It was also noted that the uglification settings were set to mangle/compress in debug mode and beautify in production mode. Presumably, this was intended to be the reverse.

### CODE REVIEW

**Comment by Timodie** [comment]
Will we need to bump up the version number as part of the node upgrade?
I noticed that the version bump followed the node upgrade in some of the other [projects](https://github.com/silvermine/toolbox/commit/a23a5dbe5ce1c8a1274aad7a226f2b88af0c8f8d)

**Comment by onebytegone** [comment]
> Will we need to bump up the version number as part of the node upgrade? I noticed that the version bump followed the node upgrade in some of the other [projects](https://github.com/silvermine/toolbox/commit/a23a5dbe5ce1c8a1274aad7a226f2b88af0c8f8d)

@Timodie Yes, once this PR is merged and if no additional changes are needed, we'll cut a new version of this package. We don't typically have a version bump in the same PR as other changes.

**Comment by joshuacurtiss** [comment]
I updated the PR description (after a closer before/after examination of the `browserslist` output after regenerating the package-lock.json) to explain that `const` being passed through Babel's output is due to `browserslist` removing `safari 9.1`, which only has partial support for `const`.

I was unsure if this change in Babel output would be acceptable, but if according to caniuse/browserslist/babel it is safe to use `const` when Safari 9.1 is eliminated, then our new built artifacts containing `const` should be alright, given that our current desired Safari support is _iOS Safari 12.2_.

CC: @kmuncie

**Comment by joshuacurtiss** [comment]
@onebytegone, would you mind giving me your thoughts on 80d826ecc3f1209dd03f52b0081492b60990a5b6? Namely:

- Incorporating the `.nvmrc` version into the test matrix
- Using `git status --porcelain` to check for uncommitted changes

I'm just checking that this particular approach is what you had in mind for accomplishing these two goals in GitHub Actions jobs.

**Comment by joshuacurtiss** [comment]
> * Using `git status --porcelain` to check for uncommitted changes

Come to think of it, there's a case for saying that the "uncommitted changes" step should be last in the `build` job, in the unlikely event that even standards checks or other tests have compromised code..

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -5,34 +5,40 @@ on: [ push, pull_request ]
 jobs:
    build:
       runs-on: ubuntu-latest
+      outputs:
+         nvmrc: ${{ steps.makeNodeVersionOutput.outputs.nvmrc }}
       steps:
-         -
-            uses: actions/checkout@v3
+         -  uses: actions/checkout@v4
```
Comment (onebytegone) [inline .github/workflows/ci.yml]: Unless it's a single key, we typically put the keys on separate lines:

```suggestion
         -
            uses: actions/checkout@v3
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -5,34 +5,40 @@ on: [ push, pull_request ]
...
+         -  uses: actions/setup-node@v4
             with:
                node-version-file: '.nvmrc'
-         - run: npm i -g npm@8.5.5
-         - run: npm ci
-         - run: npm run standards
+         -  name: Put NVM version in output
+            id: makeNodeVersionOutput
+            run: echo "nvmrc=$(cat .nvmrc)" >> "$GITHUB_OUTPUT"
+         -  run: npm ci
```
Comment (onebytegone) [inline .github/workflows/ci.yml]: For single keys, we'll typically only have one space after the dash.

```suggestion
         - run: npm ci
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -5,34 +5,40 @@ on: [ push, pull_request ]
...
-         - run: npm i -g npm@8.5.5
-         - run: npm ci # Reinstall the dependencies to ensure they install with the current version of node
-         - run: npm test
-         - name: Coveralls
-           uses: coverallsapp/github-action@v1
+         -  run: npm ci # Reinstall the dependencies to ensure they install with the current version of node
+         -  run: npm run standards
+         -  run: npm test
+         -  name: Coveralls
+            uses: coverallsapp/github-action@v2
```
Comment (onebytegone) [inline .github/workflows/ci.yml]: This should probably become a parallel job like we did in https://github.com/silvermine/toolbox/commit/011506189f1db4e2ca7c2b870fd7fc08a4ed298e

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -5,34 +5,40 @@ on: [ push, pull_request ]
...
-         - run: npm ci
-         - run: npm run standards
+         -  name: Put NVM version in output
+            id: makeNodeVersionOutput
+            run: echo "nvmrc=$(cat .nvmrc)" >> "$GITHUB_OUTPUT"
+         -  run: npm ci
+         -  name: Check for uncommitted changes # Done after dependency install and build to ensure code isn't compromised
+            run: if [ -n "$(git status --porcelain)" ]; then echo 'There are uncommitted changes.'; exit 1; fi
+         -  run: npm run check-node-version
+         -  run: npm run standards
```
Comment (onebytegone) [inline .github/workflows/ci.yml]:
```suggestion
         -  run: npm run standards
         -  run: npm run build
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -5,34 +5,40 @@ on: [ push, pull_request ]
...
             with:
                node-version: ${{ matrix.node-version }}
-         - run: npm i -g npm@8.5.5
-         - run: npm ci # Reinstall the dependencies to ensure they install with the current version of node
-         - run: npm test
-         - name: Coveralls
-           uses: coverallsapp/github-action@v1
+         -  run: npm ci # Reinstall the dependencies to ensure they install with the current version of node
+         -  run: npm run standards
+         -  run: npm test
```
Comment (onebytegone) [inline .github/workflows/ci.yml]:
```suggestion
         -  run: npm run build # Ensure building is possible with this version of node
         -  run: npm test
```

**Review by onebytegone — CHANGES_REQUESTED** [review]

**Comment by joshuacurtiss** [comment]
@onebytegone Ok, made all suggested changes.. 

Note I also moved the "uncommitted changes" check as the last step of the `build` job, since it makes more sense to be last.

Also note... to serve as an example, it makes sense to add `npm run build`. However, this particular repo has a `prepublish` script which runs a build.  Since `prepublish` runs on npm install, the project is technically already built with the `npm ci` command. Just wanted to point that out before we proceed!

**Review by onebytegone — APPROVED** [review]

**Pull request merged**